File: immersinglinker-sdk-node/src/clients/lesson.ts
import { ServiceClientBase } from './base.js';
import type {
  Subject,
  TimeState,
  TimeLayoutItem,
  ClassPlan,
  Profile,
} from '../types/lesson.js';

/** 课表/课程信息客户端，提供当前课程、上下节课、计时器、配置文件等查询接口 */
export class LessonServiceClient extends ServiceClientBase {
  // #region Current

  /** 获取当前课程科目，无课时返回 null */
  async getCurrentSubject(): Promise<Subject | null> {
    return this._getOrNull<Subject>('/lesson/current/subject');
  }

  /** 获取下一节课程的科目 */
  async getNextClassSubject(): Promise<Subject> {
    return this._get<Subject>('/lesson/current/next-class-subject');
  }

  /** 获取当前时间状态（上课/课间/无） */
  async getCurrentState(): Promise<TimeState> {
    return this._get<TimeState>('/lesson/current/state');
  }

  /** 获取当前时间布局项 */
  async getCurrentTimeLayoutItem(): Promise<TimeLayoutItem> {
    return this._get<TimeLayoutItem>('/lesson/current/time-layout-item');
  }

  /** 获取当前班级计划，无计划时返回 null */
  async getCurrentClassPlan(): Promise<ClassPlan | null> {
    return this._getOrNull<ClassPlan>('/lesson/current/class-plan');
  }

  /** 获取当前课表选中的索引 */
  async getCurrentSelectedIndex(): Promise<number> {
    return this._get<number>('/lesson/current/selected-index');
  }

  /** 获取当前班级计划是否启用 */
  async getIsClassPlanEnabled(): Promise<boolean> {
    return this._get<boolean>('/lesson/current/is-class-plan-enabled');
  }

  /** 获取当前班级计划是否已加载 */
  async getIsClassPlanLoaded(): Promise<boolean> {
    return this._get<boolean>('/lesson/current/is-class-plan-loaded');
  }

  /** 获取当前课程是否已确认 */
  async getIsLessonConfirmed(): Promise<boolean> {
    return this._get<boolean>('/lesson/current/is-lesson-confirmed');
  }

  // #endregion

  // #region Next

  /** 获取下一节课的时间布局项 */
  async getNextClassTimeLayoutItem(): Promise<TimeLayoutItem> {
    return this._get<TimeLayoutItem>('/lesson/next/class-time-layout-item');
  }

  /** 获取下一个课间的时间布局项 */
  async getNextBreakingTimeLayoutItem(): Promise<TimeLayoutItem> {
    return this._get<TimeLayoutItem>('/lesson/next/breaking-time-layout-item');
  }

  // #endregion

  // #region Previous

  /** 获取上一节课的科目，无上一节时返回 null */
  async getPreviousClassSubject(): Promise<Subject | null> {
    return this._getOrNull<Subject>('/lesson/previous/class-subject');
  }

  /** 获取上一节课的时间布局项 */
  async getPreviousClassTimeLayoutItem(): Promise<TimeLayoutItem> {
    return this._get<TimeLayoutItem>('/lesson/previous/class-time-layout-item');
  }

  /** 获取上一个课间的时间布局项 */
  async getPreviousBreakingTimeLayoutItem(): Promise<TimeLayoutItem> {
    return this._get<TimeLayoutItem>(
      '/lesson/previous/breaking-time-layout-item',
    );
  }

  // #endregion

  // #region Timer

  /** 获取当前课程剩余时间（ISO 8601 持续时间字符串） */
  async getOnClassLeftTime(): Promise<string> {
    return this._get<string>('/lesson/timer/on-class-left');
  }

  /** 获取当前课间剩余时间 */
  async getOnBreakingLeftTime(): Promise<string> {
    return this._get<string>('/lesson/timer/on-breaking-left');
  }

  /** 获取自上一节课程开始以来经过的时间 */
  async getElapsedSincePreviousClass(): Promise<string> {
    return this._get<string>('/lesson/timer/elapsed-since-previous-class');
  }

  /** 获取自上一课间开始以来经过的时间 */
  async getElapsedSincePreviousBreaking(): Promise<string> {
    return this._get<string>('/lesson/timer/elapsed-since-previous-breaking');
  }

  /** 获取自上一个任意状态开始以来经过的时间 */
  async getElapsedSincePreviousAny(): Promise<string> {
    return this._get<string>('/lesson/timer/elapsed-since-previous-any');
  }

  // #endregion

  // #region Profile

  /** 获取当前配置文件的路径 */
  async getCurrentProfilePath(): Promise<string> {
    return this._get<string>('/lesson/profile/current-profile-path');
  }

  /** 获取当前配置文件是否受信任 */
  async getIsCurrentProfileTrusted(): Promise<boolean> {
    return this._get<boolean>('/lesson/profile/is-trusted');
  }

  /** 获取当前完整配置文件 */
  async getProfile(): Promise<Profile> {
    return this._get<Profile>('/lesson/profile');
  }

  // #endregion
}

File: immersinglinker-sdk-py/immersinglinker/clients/lesson.py
from __future__ import annotations

import json
from datetime import timedelta

import aiohttp

from ..enums import TimeState
from ..types import Subject, TimeLayoutItem, ClassPlan, Profile
from .base import ImmersingLinkerError, _parse_timedelta


class LessonServiceClient:
    def __init__(self, port: str) -> None:
        self._base_url = f"http://localhost:{port}"
        self._session: aiohttp.ClientSession | None = None

    async def __aenter__(self) -> LessonServiceClient:
        self._session = aiohttp.ClientSession(self._base_url)
        return self

    async def __aexit__(self, *args: object) -> None:
        if self._session:
            await self._session.close()
            self._session = None

    async def _ensure_session(self) -> aiohttp.ClientSession:
        if self._session is None:
            raise ImmersingLinkerError(
                "Client not opened. Use 'async with LessonServiceClient(...) as client:'"
            )
        return self._session

    async def _get(
        self, path: str, *, nullable: bool = False
    ) -> str:
        session = await self._ensure_session()
        async with session.get(path) as response:
            if nullable and response.status == 404:
                return ""
            if response.status != 200:
                text = await response.text()
                raise ImmersingLinkerError(f"HTTP {response.status}: {text}")
            return await response.text()

    # region Current

    async def get_current_subject(self) -> Subject | None:
        text = await self._get("/lesson/current/subject", nullable=True)
        return Subject.from_dict(json.loads(text)) if text else None

    async def get_next_class_subject(self) -> Subject:
        text = await self._get("/lesson/current/next-class-subject")
        return Subject.from_dict(json.loads(text))

    async def get_current_state(self) -> TimeState:
        return TimeState(int(json.loads(await self._get("/lesson/current/state"))))

    async def get_current_time_layout_item(self) -> TimeLayoutItem:
        text = await self._get("/lesson/current/time-layout-item")
        return TimeLayoutItem.from_dict(json.loads(text))

    async def get_current_class_plan(self) -> ClassPlan | None:
        text = await self._get("/lesson/current/class-plan", nullable=True)
        return ClassPlan.from_dict(json.loads(text)) if text else None

    async def get_current_selected_index(self) -> int:
        return int(await self._get("/lesson/current/selected-index"))

    async def get_is_class_plan_enabled(self) -> bool:
        raw = await self._get("/lesson/current/is-class-plan-enabled")
        value = json.loads(raw)
        if isinstance(value, bool):
            return value
        return str(value).strip().lower() == "true"

    async def get_is_class_plan_loaded(self) -> bool:
        raw = await self._get("/lesson/current/is-class-plan-loaded")
        value = json.loads(raw)
        if isinstance(value, bool):
            return value
        return str(value).strip().lower() == "true"

    async def get_is_lesson_confirmed(self) -> bool:
        raw = await self._get("/lesson/current/is-lesson-confirmed")
        value = json.loads(raw)
        if isinstance(value, bool):
            return value
        return str(value).strip().lower() == "true"

    # endregion

    # region Next

    async def get_next_class_time_layout_item(self) -> TimeLayoutItem:
        text = await self._get("/lesson/next/class-time-layout-item")
        return TimeLayoutItem.from_dict(json.loads(text))

    async def get_next_breaking_time_layout_item(self) -> TimeLayoutItem:
        text = await self._get("/lesson/next/breaking-time-layout-item")
        return TimeLayoutItem.from_dict(json.loads(text))

    # endregion

    # region Previous

    async def get_previous_class_subject(self) -> Subject | None:
        text = await self._get("/lesson/previous/class-subject", nullable=True)
        return Subject.from_dict(json.loads(text)) if text else None

    async def get_previous_class_time_layout_item(self) -> TimeLayoutItem:
        text = await self._get("/lesson/previous/class-time-layout-item")
        return TimeLayoutItem.from_dict(json.loads(text))

    async def get_previous_breaking_time_layout_item(self) -> TimeLayoutItem:
        text = await self._get("/lesson/previous/breaking-time-layout-item")
        return TimeLayoutItem.from_dict(json.loads(text))

    # endregion

    # region Timer

    async def get_on_class_left_time(self) -> timedelta:
        return _parse_timedelta(await self._get("/lesson/timer/on-class-left"))

    async def get_on_breaking_left_time(self) -> timedelta:
        return _parse_timedelta(await self._get("/lesson/timer/on-breaking-left"))

    async def get_elapsed_since_previous_class(self) -> timedelta:
        return _parse_timedelta(await self._get("/lesson/timer/elapsed-since-previous-class"))

    async def get_elapsed_since_previous_breaking(self) -> timedelta:
        return _parse_timedelta(await self._get("/lesson/timer/elapsed-since-previous-breaking"))

    async def get_elapsed_since_previous_any(self) -> timedelta:
        return _parse_timedelta(await self._get("/lesson/timer/elapsed-since-previous-any"))

    # endregion

    # region Profile

    async def get_current_profile_path(self) -> str:
        raw = await self._get("/lesson/profile/current-profile-path")
        return json.loads(raw)

    async def get_is_current_profile_trusted(self) -> bool:
        raw = await self._get("/lesson/profile/is-trusted")
        value = json.loads(raw)
        if isinstance(value, bool):
            return value
        return str(value).strip().lower() == "true"

    async def get_profile(self) -> Profile:
        text = await self._get("/lesson/profile")
        return Profile.from_dict(json.loads(text))

    # endregion

    async def close(self) -> None:
        if self._session:
            await self._session.close()
            self._session = None
